Source/Dafny/Verifier/Translator.cs

```diff
@@ -6260,7 +6260,37 @@ private class WFOptions {
 
       } else if (expr is StmtExpr) {
         var e = (StmtExpr)expr;
+        // If we're inside an "old" expression, then "etran" will know how to translate
+        // expressions. However, here, we're also having to translate e.S, which is a
+        // Statement. Since statement translation (in particular, CallStmt's) works
+        // directly on the global variable $Heap, we change its value here.
+        // (Note, this would be a bad translation if the Statement could introduce
+        // control flow that would miss the "$Heap := tmpHeap;" assignment that
+        // resets the value of $Heap. Luckily, StmtExpr does not allow any such
+        // Statement.)
+        Bpl.IdentifierExpr tmpHeap = null;
+        Bpl.IdentifierExpr theHeap = null;
+        if (etran.UsesOldHeap) {
+          var suffix = CurrentIdGenerator.FreshId("StmtExpr#");
+          var tmpHeapVar = new Bpl.LocalVariable(e.S.Tok, new Bpl.TypedIdent(e.S.Tok, "Heap$" + suffix, predef.HeapType));
+          locals.Add(tmpHeapVar);
+          tmpHeap = new Bpl.IdentifierExpr(e.S.Tok, tmpHeapVar);
+          var generalEtran = new ExpressionTranslator(this, predef, e.S.Tok);
+          theHeap = (Bpl.IdentifierExpr /*TODO: this cast is dubious*/)generalEtran.HeapExpr;
```

Comment:
Please resolve the TODO

Reply:
Such a cast is done in many places in the translator. Each one is dubious (but, apparently, we are not seeing any examples where these casts actually fail). I could refactor all occurrences of this pattern to one place, but that one place would still have the comment. Would that be an improvement? Can you think of some better solution?

Reply:
A TODO implies something has to be done, but if you merge it, then the moment when TODO that is not clear, and I'd say at that point it's never. If you want to allocate time to resolving this, I think creating a GH issue will at least prompt someone to consider picking it up, or deciding not to and closing the issue.

If you don't want to allocate time, but just help future developers read the code, I'd suggest changing this into a comment, `// we don't know why but this cast never fails in practice`, and then indeed it would help centralising the cast.